File: MCTabBarDemo/MCTabBar/MCTabBarController.m
//
//  MCTabBarController.m
//  MCTabBarDemo
//
//  Created by chh on 2017/12/4.
//  Copyright © 2017年 Mr.C. All rights reserved.
//

#import "MCTabBarController.h"
#import "ViewController.h"

@interface MCTabBarController ()<UITabBarControllerDelegate>
@property (nonatomic, strong) UIButton *button;
@end

@implementation MCTabBarController

- (void)viewDidLoad {
    [super viewDidLoad];
    UITabBar *tabbar = [UITabBar appearance];
    tabbar.tintColor = [UIColor colorWithRed:27.0/255.0 green:118.0/255.0 blue:208/255.0 alpha:1];//选中时的颜色
    //设置为NO后则没有透明度添加(tabbar为白色)
    tabbar.translucent = NO;
    self.delegate = self;
    [self addChildViewControllers];
    [self setUpButton];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

//添加子控制器
- (void)addChildViewControllers{
    //图片大小建议32*32
    [self addChildrenViewController:[[ViewController alloc] init] andTitle:@"首页" andImageName:@"tab1_n" andSelectImage:@"tab1_p"];
    [self addChildrenViewController:[[ViewController alloc] init] andTitle:@"扩展" andImageName:@"tab2_n" andSelectImage:@"tab2_p"];
    //中间这个不设置东西，只占位
    [self addChildrenViewController:[[ViewController alloc] init] andTitle:@"旋转" andImageName:@"" andSelectImage:@""];
    [self addChildrenViewController:[[ViewController alloc] init] andTitle:@"发现" andImageName:@"tab3_n" andSelectImage:@"tab3_p"];
    [self addChildrenViewController:[[ViewController alloc] init] andTitle:@"我" andImageName:@"tab4_n" andSelectImage:@"tab4_p"];
}

- (void)addChildrenViewController:(UIViewController *)childVC andTitle:(NSString *)title andImageName:(NSString *)imageName andSelectImage:(NSString *)selectedImage{
    childVC.tabBarItem.image = [UIImage imageNamed:imageName];
    childVC.tabBarItem.selectedImage =  [UIImage imageNamed:selectedImage];
    childVC.title = title;
    
    UINavigationController *baseNav = [[UINavigationController alloc] initWithRootViewController:childVC];
    
    [self addChildViewController:baseNav];
}

//中间按钮布局
- (void)setUpButton{
    _button = [UIButton buttonWithType:UIButtonTypeCustom];
    //  设定button大小为适应图片
    UIImage *normalImage = [UIImage imageNamed:@"tabbar_add"];
    _button.frame = CGRectMake(0, 0, normalImage.size.width, normalImage.size.height);
    [_button setImage:normalImage forState:UIControlStateNormal];
    [_button addTarget:self action:@selector(buttonAction:) forControlEvents:UIControlEventTouchUpInside];
    //去除选择时高亮
    _button.adjustsImageWhenHighlighted = NO;
    //根据图片调整button的位置
    CGPoint center = self.tabBar.center;
    center.y = self.tabBar.frame.origin.y + 5;//中心为tabbar顶部
    _button.center = center;
    [self.view addSubview:_button];
}

- (void)buttonAction:(UIButton *)button{
    self.selectedIndex = 2;//关联中间按钮
    [self rotationAnimation];
}

//tabbar选择时的代理
- (void)tabBarController:(UITabBarController *)tabBarController didSelectViewController:(UIViewController *)viewController{
    if (tabBarController.selectedIndex == 2){//选中中间的按钮
        [self rotationAnimation];
    }else {
        [_button.layer removeAllAnimations];
    }
}
//旋转动画
- (void)rotationAnimation{
    CABasicAnimation *rotationAnimation = [CABasicAnimation animationWithKeyPath:@"transform.rotation.z"];
    rotationAnimation.toValue = [NSNumber numberWithFloat:M_PI*2.0];
    rotationAnimation.duration = 3.0;
    rotationAnimation.repeatCount = HUGE;
    [_button.layer addAnimation:rotationAnimation forKey:@"key"];
}

@end

File: MCTabbarDemoSwift/MCTabbar/MCTabbarController.swift
//
//  MCTabbarController.swift
//  MCTabbarDemoSwift
//
//  Created by caohouhong on 2018/12/4.
//  Copyright © 2018年 caohouhong. All rights reserved.
//

import UIKit

class MCTabbarController: UITabBarController,UITabBarControllerDelegate{
    
    var mTabBar: MCTabBar!
    var selectItem: Int = 0 // 选中的item
    
    override func viewDidLoad() {
        super.viewDidLoad()
        mTabBar = MCTabBar(frame: CGRect.zero)
        mTabBar.centerBtn.addTarget(self, action: #selector(centerBtnAction), for: .touchUpInside)
        //选中时的颜色
        mTabBar.tintColor = UIColor(red: 27.0/255.0, green: 118.0/255.0, blue: 208/255.0, alpha: 1)
        //透明设置为NO，显示白色，view的高度到tabbar顶部截止，YES的话到底部
        mTabBar.isTranslucent = false
        self.setValue(mTabBar, forKeyPath: "tabBar")
        self.delegate = self as UITabBarControllerDelegate
       
        addChildViewControllers()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        
    }
    
    //添加子控制器
    func addChildViewControllers() {
        //图片大小建议32*32
        addChildViewController(ViewController(), title: "首页", imageName: "tab1")
        addChildViewController(ViewController(), title: "扩展", imageName: "tab2")
        //中间这个不设置东西，只占位
        addChildViewController(ViewController(), title: "旋转", imageName: "")
        addChildViewController(ViewController(), title: "发现", imageName: "tab3")
        addChildViewController(ViewController(), title: "我的", imageName: "tab4")
    }
    
    // 添加子vc
    func addChildViewController(_ childController: UIViewController, title: String, imageName: String) {
        childController.title = title
        childController.tabBarItem.image = UIImage(named: imageName)
        childController.tabBarItem.selectedImage = UIImage(named: imageName)
        let nav = BaseNavigationController(rootViewController: childController)
        addChildViewController(nav)
    }
    
    //tabbar选择时的代理
    func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController) {
        if tabBarController.selectedIndex == 2 {//选中中间的按钮
            if (self.selectItem != tabBarController.selectedIndex){// 如果是重复选择不做处理
                rotationAnimation()
            }
        }else {
            mTabBar.centerBtn.layer.removeAllAnimations()
        }
        self.selectItem = tabBarController.selectedIndex
    }
    
    // 中间按钮点击
    @objc func centerBtnAction() {
        self.selectedIndex = 2// 关联中间按钮
        if (self.selectItem != self.selectedIndex){// 重复选择不做处理
            rotationAnimation()
        }
        self.selectItem = self.selectedIndex
    }
    
    // 旋转动画
    func rotationAnimation() {
        let animation = CABasicAnimation(keyPath: "transform.rotation.z")
        animation.toValue = NSNumber(value: Double.pi*2.0)
        animation.duration = 3.0
        animation.repeatCount = HUGE
        mTabBar.centerBtn.layer.add(animation, forKey: "animation")
    }
}
